text/0000-grid-primitives.md

```diff
@@ -0,0 +1,760 @@
+    production_org='philipsbulbfactory',
+    properties=[
+        PropertyValue(
+            name="size",
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_value=10,
+        ),
+        PropertyValue(
+            name="bulb_type"
+            data_type=PropertyDefinition.DataType.ENUM,
+            enum_value=2,  # LED
+        ),
+        PropertyValue(
+            name="energy_rating",
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_value=89,
+        ),
+        PropertyValue(
+            name="color",
+            data_type=PropertyDefinition.DataType.STRUCT,
+            struct_values=[
+                PropertyValue(
+                    name="name",
+                    data_type=PropertyDefinition.DataType.STRING,
+                    string_value="White",
+                ),
+                PropertyValue(
+                    name="rgb_hex",
+                    data_type=PropertyDefinition.DataType.STRING,
+                    string_value="000000",
+                ),
+           ]),
+    ])
+```
+
+## Addressing
+
+### Referencing Schemas
+
+Schemas are uniquely referenced by their unique schema name. For example:
+
+```
+    get_schema(schema_name)
+    set_schema(schema_name, schema)
+```
+
+### Schema Addressing in the Merkle-Radix State System
+
+Grid Schemas are stored under the Grid namespace 621dee. For each schema, the
+address is formed by concatenating the namespace, the special policy namespace
+of 01, and the first 62 characters of the SHA-256 hash of the schema name.
+
+For example, the address of the "Lightbulb" schema defined in the example above
+is (in Python): 
+
+```
+    "621dee" + "01" + hashlib.sha512("Lightbulb").encode("utf-8")).hexdigest()[:62]
+```
+
+To avoid hash collisions, schemas must be stored in a `SchemaList`.
+
+## Transactions
+
+In order to add Schemas to state, a transaction must be used.  The following
+transactions and their execution rules are designed for the Hyperledger Sawtooth
+platform and may differ for other transaction execution platforms.
+
+### Transaction Header
+
+The header for the transactions will include the following:
+
+- `family_name`: `"grid_schema"`
+- `family_version`: `"1.0"`
+- `namespaces`: `[ "621dee" ]`
+
+### Payloads and Execution Rules
+
+#### SchemaPayload
+
+SchemaPayload contains an action enum and the associated action payload.  This
+allows for the action payload to be dispatched to the appropriate logic.
+
+Only the defined actions are available and only one action payload should be
+defined in the SchemaPayload.
+
+```
+message SchemaPayload {
+    enum Actions {
+        UNSET_ACTION = 0;
+        SCHEMA_CREATE = 1;
+        SCHEMA_UPDATE = 2;
+    }
+
+    Action action = 1;
+
+    SchemaCreateAction schema_create = 2;
+    SchemaUpdateAction schema_update = 3
+}
+```
+
+#### SchemaCreateAction
+
+SchemaCreateAction adds a new Schema to state. 
+
+```
+message SchemaCreateAction {
```

Comment:
Shouldn't this and the Update action take a `SchemaList`? Or is the contract going to handle that wrap/unwrap transparently?  If so, feels like emphasizing that need to wrap in list may not be necessary?

Reply:
`SchemaList` is mainly for handling address collisions - think of it like a bucket in a hash table.  The create and update transaction operate on a single schema at a time.  It's up to the smart contract implementation to handle the collisions, when they occur.

Reply:
Yeah that was my assumption as well, but the language in the doc above says "Schemas must be stored in a Schema List".  If this is really an implementation nuance of the smart contract and not something that would ever be "handled" it's maybe less of a fundamental Grid primitive.  But probably fine either way, no real strong opinions here.